monitoring/nethealth.go

```diff
@@ -0,0 +1,411 @@
+				Warn("Timeout counter increased more than request counter.")
+			continue
+		}
+
+		// Record the new packet loss percentage, but only keep the last `seriesCapacity` number of data points.
+		if len(data.packetLossPercentages) >= c.seriesCapacity {
+			data.packetLossPercentages = data.packetLossPercentages[1:]
+		}
+		packetLoss := timeoutInc / requestInc
+		data.packetLossPercentages = append(data.packetLossPercentages, packetLoss)
+
+		if err := c.peerStats.Set(peer, data); err != nil {
+			return updated, trace.Wrap(err)
+		}
+
+		// Record updated peers to be returned for later use in network verification step.
+		updated = append(updated, peer)
+	}
+	return updated, nil
+}
+
+// verifyNethealth verifies that the overlay network communication is healthy
+// for the nodes specified by the provided list of peers. Failed probes will be
+// reported for unhealthy peers.
+func (c *nethealthChecker) verifyNethealth(peers []string) (health.Reporter, error) {
+	reporter := new(health.Probes)
+	for _, peer := range peers {
+		healthy, err := c.isHealthy(peer)
+		if err != nil {
+			return reporter, trace.Wrap(err)
+		}
+		if !healthy {
+			reporter.Add(NewProbeFromErr(c.Name(), nethealthDetail(peer), nil))
+		}
+	}
+	return reporter, nil
+}
+
+// isHealthy returns true if the overlay network is healthy for the specified
+// peer.
+func (c *nethealthChecker) isHealthy(peer string) (healthy bool, err error) {
+	packetLossPercentages, err := c.peerStats.GetPacketLoss(peer)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+
+	// Checker has not collected enough data yet to check network health.
+	if len(packetLossPercentages) < c.seriesCapacity {
+		return true, nil
+	}
+
+	// If the percentage is above `packetLossTreshhold` throughout the entire
+	// interval, the overlay network communication to that peer will be
+	// considered unhealthy.
+	for i := 0; i < c.seriesCapacity; i++ {
+		if packetLossPercentages[i] <= packetLossThreshold {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// parseCounter parses the provided metricFamilies and returns a map of counters
+// for the desired metrics specified by label.
+func parseCounter(metricFamilies map[string]*dto.MetricFamily, label string) (counters map[string]float64, err error) {
+	mf, ok := metricFamilies[label]
+	if !ok {
+		return nil, trace.NotFound("%s metrics not found", label)
+	}
+
+	counters = make(map[string]float64)
+	for _, m := range mf.GetMetric() {
+		peerName, err := getPeerName(m.GetLabel())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		counters[peerName] = m.GetCounter().GetValue()
+	}
+	return counters, nil
+}
+
+// getPeerName extracts the 'peer_name' value from the provided labels.
+func getPeerName(labels []*dto.LabelPair) (peer string, err error) {
+	for _, label := range labels {
+		if peerLabel == label.GetName() {
+			return label.GetValue(), nil
+		}
+	}
+	return "", trace.NotFound("unable to find required peer label")
```

Comment:
Error could be a bit better, like which label?

Reply:
Modified to print out `peer_name` label.